### PR TITLE
Prepare 1.6.0 release

### LINKED ISSUE #2424: January release 1.6.0.0
The machine never stops! Time to plan the new release.

* will include two new brand plugins
* and the continuous work of @pepeiborra improving performance
* [x] it will remove support for ghc-8.10.5
* [x] it should include the artifacts from gitlab, hopefully in a automated way (see https://github.com/haskell/ghcup-metadata/pull/11 and #2420
* [x] as noted by @hasufell the version of the gh release should match the hackage one (so 1.6.0.0)

as always if anyone feels strong enough to drive the release or collaborate with, help will be very welcomed :smiley_cat: 

-----------------------------

### prerelease sanity checks

- [x] trigger manually the hackage workflow *without* uploading the packages: https://github.com/jneira/haskell-language-server/actions/runs/1751828521
- [X] trigger manually the build workflow : https://github.com/jneira/haskell-language-server/actions/runs/1753316308
- [x] create a prerelease tag `1.6.0-check-gitlab` and push it to the [project repo in gitlab](https://gitlab.haskell.org/haskell/haskell-language-server) to check the build is fine: https://gitlab.haskell.org/haskell/haskell-language-server/-/pipelines/46483

### github release

- [x] generate the list of pull requests finished since the last release using the [haskell script](https://github.com/haskell/haskell-language-server/blob/master/GenChangelogs.hs) in the project root.
  Nix users should run command `gen-hls-changelogs` (a wrapper of the script) in nix-shell instead.
- [x] add that list to the actual [Changelog](https://github.com/haskell/haskell-language-server/blob/master/ChangeLog.md) with a description of the release.
- [x] bump up versions of changed packages. All are optional but [haskell-language-server itself](https://github.com/haskell/haskell-language-server/blob/master/haskell-language-server.cabal).
- [x] create the tag and make an initial prerelease to trigger the ci workflow (see details below)
- [x] contact ghcup team (#haskell-ghcup irc channel or via its [repo](https://github.com/haskell/ghcup-metadata)) to try to sync our release and its inclusion in ghcup
- [x] in the github release edit page, check the attached binaries and the release description (usually the changelog entry) and uncheck the prerelease box
- [ ] make public the release in the usual social channels (not required but useful to spread the word :slightly_smiling_face:):
  - [x] irc
  - [x] matrix
  - [x] twitter
  - [x] discord
  - [x] discourse
  - [x] reddit

### hackage release

- [x] bump up package versions following the [pvp specification](https://pvp.haskell.org/) if they are not already updated. You could use [policeman](https://github.com/kowainik/policeman) to help with this step.
- [x] create 1.6.0-hackage branch to trigger the hackage github workflow which will upload all changed packages to hackage as candidates
- [ ] check manually candidates in hackage
- [ ] publish them definitely

### ghcup release

- [x] push the release tag to the [haskell-language-server gitlab repo](https://gitlab.haskell.org/haskell/haskell-language-server) to trigger the build of ghcup specific artifacts
- [x] download specific artifacts only available in the gitlab build and compute their sha256sum
- [x] upload them to the github release and complete the SHA256SUMS file
- [x] change ghcup metadata to include the new release in https://github.com/haskell/ghcup-metadata
  -  example pull request [here](https://github.com/haskell/ghcup-metadata/pull/11)
  - pull request: https://github.com/haskell/ghcup-metadata/pull/19

### PR DESCRIPTION
**Pull request by jneira**
* Closes #2424

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/2642"></a>

### CODE REVIEW

File: .github/workflows/hackage.yml
Comment (michaelpj) [inline .github/workflows/hackage.yml:110]: big hammer?

Comment (jneira) [reply .github/workflows/hackage.yml:110]: nope, below there are the concrete allow newer

Comment (jneira) [reply .github/workflows/hackage.yml:110]: nope, just below there are the concrete allow newers needed to make the solver happy, the section will look:

```
allow-newer:
 butcher:base, multistate:base, data-tree-print:base,
 stylish-haskell:Cabal,stylish-haskell:ghc-lib-parser,stylish-haskell:aeson
```

Comment (jneira) [reply .github/workflows/hackage.yml:110]: So brittany cant be built from hackage and ghc-9.0.2, will update the hls.cabal file if needed

**Comment by michaelpj** [comment]
I guess we probably need to bump the versions of a bunch of plugins too?

**Comment by jneira** [comment]
> I guess we probably need to bump the versions of a bunch of plugins too?

Yeah, i was just doing that. It is enforced by the hackage workflow: it fails if the local package being tested differs from the hackage version but has the same cabal version

File: ChangeLog.md
Comment (jhrcek) [inline ChangeLog.md:20]:
```suggestion
  - We recommend upgrading ghc to the last minor version: 8.8.4 or 8.10.7
```

File: ChangeLog.md
Comment (jhrcek) [inline ChangeLog.md:25]:
```suggestion
    - If you want to install hls from source we recommend using `ghcup`. Download it and run `ghcup compile hls --help` to get more info about.
```

File: docs/supported-versions.md
Comment (jhrcek) [inline docs/supported-versions.md:31]:
```suggestion
Sometimes a plugin can be supported in prebuilt binaries but not in a hls binary installed from hackage.
```

**Review by jhrcek — COMMENTED** [review]
Few stylistic suggestions..

File: ChangeLog.md
Comment (pepeiborra) [inline ChangeLog.md:13]:
```suggestion
  - Thanks to a great collective effort coordinated by @pepeiborra and with the help of @wz1000, @mpickering and @alanz among others
```

File: ChangeLog.md
Comment (pepeiborra) [inline ChangeLog.md:203]: Since we have rolled back the Musl change, we should tweak this warning to fire not only in Mac but also in Linux

**Review by pepeiborra — CHANGES_REQUESTED** [review]

**Review by fendor — APPROVED** [review]
LGTM, thanks for taking care of the release!

File: ChangeLog.md
Comment (jneira) [reply ChangeLog.md:203]: agree, will do in this pr

**Comment by jneira** [comment]
@jhrcek thanks for corrections 🙂

File: ChangeLog.md
Comment (Ailrun) [inline ChangeLog.md:27]: To make this more clear, I think it would be better to add a deprecation note at the beginning of [this doc](https://haskell-language-server.readthedocs.io/en/latest/installation.html#install-specific-ghc-version).
If a user scrolls down to this next section, he/she may get confused.

Comment (jneira) [reply ChangeLog.md:27]: hmm, the plan would be replace installation instructions removing all references to install script. I thought do it just after the release but you are right if someone follows the link in the meanwhile it will see contradictory assertions

File: ChangeLog.md
Comment (jneira) [reply ChangeLog.md:203]: done with 539f00, i've relaxed the warning as th evaluation works in some cases for linux

File: .github/actions/setup-build/action.yml
Comment (eddiemundo) [inline .github/actions/setup-build/action.yml:19]:
```suggestion
    description: "Whether we have to shorten haskell language server to hls"
```

File: .github/workflows/hackage.yml
Comment (eddiemundo) [inline .github/workflows/hackage.yml:23]: Not sure what any of this does but just making sure `!=` is supposed to change to `==`?

**Review by eddiemundo — APPROVED** [review]

File: ghcide/src/Development/IDE/Core/Rules.hs
Comment (pepeiborra) [inline ghcide/src/Development/IDE/Core/Rules.hs:832]: I'm not sure about the wording changes, it seems to me that it will generate more confusion than it's worth. 

Just looking at the last couple of days in the issue tracker to see how recurrent the problem is:

* https://github.com/haskell/haskell-language-server/issues/2596
* https://github.com/haskell/haskell-language-server/issues/2054
* https://github.com/haskell/haskell-language-server/issues/2340
* https://github.com/haskell/haskell-language-server/issues/1497
* https://github.com/haskell/haskell-language-server/issues/2633

File: ChangeLog.md
Comment (jneira) [reply ChangeLog.md]: i started to write the new install doc but i dont get it right easily so i think is better to remove the reference for now

File: ghcide/src/Development/IDE/Core/Rules.hs
Comment (jneira) [reply ghcide/src/Development/IDE/Core/Rules.hs:832]: Well, the fact it *works* for some cases.
Not sure in how many cases is working cause happy users dont open issues but maybe more than we expected.
We are gonna lie to those users and make them install from source when it would not be necessary.

Maybe we can rewrite it to be precise, recommending strongly building from source without ignoring it could work in some environments

File: .github/workflows/hackage.yml
Comment (jneira) [reply .github/workflows/hackage.yml:23]: Thanks for noting it. Yeah it was wrong and this fixes it.

**Review by drsooch — APPROVED** [review]

**Review by kokobd — APPROVED** [review]

File: ChangeLog.md
Comment (Ailrun) [inline ChangeLog.md:14]: As stylish-haskell remains as a "special" case, maybe it is worth noting that the support is blocked by an upstream issue? Or even better, with a link to the issue: https://github.com/haskell/stylish-haskell/issues/378

Comment (Ailrun) [reply ChangeLog.md:14]: Also,
```suggestion
- Hls now also [supports *ghc 9.0.2*](https://github.com/haskell/haskell-language-server/issues/297) with all plugins but the stylish-haskell formatter
```
the repetition of "support" sounds a bit redundant.

**Review by Ailrun — APPROVED** [review]
Thank you, always, for all your effort to cut releases. Here is one minor comment

File: ghcide/src/Development/IDE/Core/Rules.hs
Comment (jneira) [reply ghcide/src/Development/IDE/Core/Rules.hs:832]: Ok i changed the first verb from `might`to `could` to denote the binary *likely* will not work with th:
```
"This HLS binary could not support Template Haskell."
          , "Follow this [instructions](" <> templateHaskellInstructions <> ")"
          , "to build an HLS binary with reliable support for Template Haskell."
```

File: ghcide/ghcide.cabal
Comment (michaelpj) [inline ghcide/ghcide.cabal:63]: Any reason not to drop the old constraints in the places where you've done this?

File: ghcide/src/Development/IDE/Core/Rules.hs
Comment (michaelpj) [inline ghcide/src/Development/IDE/Core/Rules.hs:831]: I think the old version read better to me.

**Review by michaelpj — APPROVED** [review]

File: .github/actions/setup-build/action.yml
Comment (jneira) [reply .github/actions/setup-build/action.yml:72]: And would simplify ci, and there would not be problems with long paths in windows, the name would be consistent with ghcup, etc, etc
I knoe it is not constructive but i cant resist to remember i suggested hls as executable name when all this started (like the executable for haskell-ide-engine was hie)
But now it would break the actual usage and who knows how many scripts and configurations out there so maybe it is too late.
If we decide to do the change it should have to be done very very carefully and gradually, supporting both names in the meanwhile 😓

File: docs/features.md
Comment (jneira) [reply docs/features.md:254]: yeah, i apply several changes suggested by markdown-lint here and duplicate headers is one of its warnings

File: .github/actions/setup-build/action.yml
Comment (drsooch) [reply .github/actions/setup-build/action.yml:72]: A simple symlink should resolve most of the script/config issues no?

File: ghcide/src/Development/IDE/Core/Rules.hs
Comment (pepeiborra) [reply ghcide/src/Development/IDE/Core/Rules.hs]: I still think that `could` is still confusing, to be honest. 

Taking a step back, your main issue here is that you don't want users to build HLS from source unnecessarily. Why is that so bad? One obvious problem is that `cabal install` will place the binary under `~/.cabal/bin` and will not annotate it with the ghc version number, which means that the vscode extension will not find it. 

So the problem is that instructing users to build from source leads to a bad experience. Should we focus on addressing that, rather than on this wording?

File: ghcide/src/Development/IDE/Core/Rules.hs
Comment (jneira) [reply ghcide/src/Development/IDE/Core/Rules.hs:831]: the does -> could part? or the -> this? the first one is a semantic change cause th could work for some cases

File: ghcide/ghcide.cabal
Comment (jneira) [reply ghcide/ghcide.cabal:63]: No really, i changed all packages to use the same hls-graph, hls-plugin-api and ghcide bounds in all of them and make updates easier. The bounds includes the current version and the previous one but i did not any test if the prev version works or does not (or if the package works with older versions)

File: ghcide/src/Development/IDE/Core/Rules.hs
Comment (michaelpj) [reply ghcide/src/Development/IDE/Core/Rules.hs:831]: Then I'd use "might" there

File: ghcide/ghcide.cabal
Comment (michaelpj) [reply ghcide/ghcide.cabal:63]: AFAICT, the only thing we can make promises about is "the current state of the repo works", so I think including the previous versions isn't reliable (and does it benefit anyone?).

Comment (jneira) [reply ghcide/ghcide.cabal:63]: agree will keep only the lastest one

File: ghcide/src/Development/IDE/Core/Rules.hs
Comment (jneira) [reply ghcide/src/Development/IDE/Core/Rules.hs:831]: I used might but I read somewhere that might denotes less probability and after @pepeiborra concern I wanted to make stronger the fact it could not work

File: .github/workflows/hackage.yml
Comment (jneira) [reply .github/workflows/hackage.yml:111]: No, it is a new cabal.project created on the fly. I think i broke it in a previous pr and i detected the error here

File: .github/actions/setup-build/action.yml
Comment (drsooch) [reply .github/actions/setup-build/action.yml:72]: Yeah I don't know all the gory details with windows as I don't use it myself. Then the downstream itself I'm not sure how ghcup (?) would handle the symlinks and if that itself breaks anything? Seems like something to push down the road.

File: ghcide/src/Development/IDE/Core/Rules.hs
Comment (pepeiborra) [reply ghcide/src/Development/IDE/Core/Rules.hs]: Since statically linking glibc is unsupported, I don't know a good way of making the warning more precise.

File: .github/actions/setup-build/action.yml
Comment (jneira) [reply .github/actions/setup-build/action.yml:72]: > Then the downstream itself I'm not sure how ghcup (?) would handle the symlinks and if that itself breaks anything?

Using tiny executable wrappers called shims

File: ghcide/src/Development/IDE/Core/Rules.hs
Comment (jneira) [reply ghcide/src/Development/IDE/Core/Rules.hs]: I am afraid my knowledge of those low level details is very limited so i cant help much.
If we cant be precise with the warning i still think we should be cautious with the message.
Users with a working th installation watching that message will lose a little bit of confidence in our error handling.

But well, as you still think the original wording is better will restore it.

File: ghcide/src/Development/IDE/Core/Rules.hs
Comment (jneira) [reply ghcide/src/Development/IDE/Core/Rules.hs]: original version restored

**Comment by jneira** [comment]
@pepeiborra would be this ready to go?

**Review by pepeiborra — APPROVED** [review]

**Comment by pepeiborra** [comment]
Thanks @jneira for cutting a new release!

**Comment by jneira** [comment]
Hi all, i firmly think that the next release *must* be driven by other maintainer. It will be good for the project as other maintainers than me should be able to do the relase. I hope the documentation about would be complete enough to do it.
Of cource will help in the process as much as i can.
Thanks!

**Pull request merged**
